src/core/lib/iomgr/tcp_posix.c

```diff
@@ -108,6 +110,29 @@ typedef struct {
   grpc_resource_user_slice_allocator slice_allocator;
 } grpc_tcp;
 
+static void add_to_estimate(grpc_tcp *tcp, size_t bytes) {
+  tcp->bytes_read_this_round += bytes;
+}
+
+static void finish_estimate(grpc_tcp *tcp) {
+  if (tcp->bytes_read_this_round > tcp->target_length) {
+    tcp->target_length = 1.5 * tcp->bytes_read_this_round;
```

Comment:
I think it was a little mal-adjusted... the intent (which is expressed better, and commented now) is to avoid the target length from directly controlling bytes read and getting stuck.